src/storage-controller/src/history.rs

```diff
@@ -0,0 +1,187 @@
+    /// The sequence of commands that should be applied.
+    ///
+    /// This list may not be "compact" in that there can be commands that could be optimized or
+    /// removed given the context of other commands, for example compaction commands that can be
+    /// unified, or run commands that can be dropped due to allowed compaction.
+    commands: Vec<StorageCommand<T>>,
+    /// Tracked metrics.
+    metrics: HistoryMetrics,
+}
+
+impl<T: std::fmt::Debug> CommandHistory<T> {
+    /// Constructs a new command history.
+    pub fn new(metrics: HistoryMetrics) -> Self {
+        metrics.reset();
+
+        Self {
+            reduced_count: 0,
+            commands: Vec::new(),
+            metrics,
+        }
+    }
+
+    /// Returns an iterator over the contained storage commands.
+    pub fn iter(&self) -> impl Iterator<Item = &StorageCommand<T>> {
+        self.commands.iter()
+    }
+
+    /// Adds a command to the history.
+    ///
+    /// This action will reduce the history every time it doubles.
+    pub fn push(&mut self, command: StorageCommand<T>) {
+        use StorageCommand::*;
+
+        self.commands.push(command);
+
+        if self.commands.len() > 2 * self.reduced_count {
+            self.reduce();
+        } else {
+            // Refresh reported metrics. `reduce` already refreshes metrics, so we only need to do
+            // that here in the non-reduce case.
+            let command = self.commands.last().expect("pushed above");
+            let metrics = &self.metrics;
+            match command {
+                CreateTimely { .. } => metrics.create_timely_count.inc(),
+                InitializationComplete => metrics.initialization_complete_count.inc(),
+                UpdateConfiguration(_) => metrics.update_configuration_count.inc(),
+                RunIngestions(x) => metrics.run_ingestions_count.add(x.len().cast_into()),
+                RunSinks(x) => metrics.run_sinks_count.add(x.len().cast_into()),
+                AllowCompaction(x) => metrics.allow_compaction_count.add(x.len().cast_into()),
+            }
+        }
+    }
+
+    /// Reduces the command history to a minimal form.
+    pub fn reduce(&mut self) {
+        use StorageCommand::*;
+
+        let mut create_timely_command = None;
+        let mut initialization_complete = false;
+        let mut final_compactions = BTreeMap::new();
+
+        // Collect the final definitions of ingestions and sinks.
+        // The same object ID can occur in multiple run commands when an object was altered. In
+        // this scenario, we only want to send the most recent definition of the object.
+        let mut final_ingestions = BTreeMap::new();
+        let mut final_sinks = BTreeMap::new();
+
+        // Collect only the final configuration.
+        // Note that this is only correct as long as all config parameters apply globally. If we
+        // ever introduce parameters that only affect subsequent commands, we will have to
```

Comment:
Yeah, this can be worded better. The danger is that it's somehow important for stability/correctness that an updated configuration only applies to the dataflows created in the command stream, so when we replay the command stream it's important that we _don't_ apply the updated config to dataflows that were created prior in the command stream. Config options that apply only to subsequent dataflows are fine, as long as we don't mind that they will get applied to all dataflows on new replicas, or reconnecting replicas.

I think another way of expressing this is that we can't have config options that change the observable output of dataflows. Options that change their performance behavior are fine.

Reply:
I changed the wording. I find it hard to find a crisp spec for what kind of options are fine and which are not. That's probably something we will need to decide on a case-by-case basis.